url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/ec/util/MersenneTwisterFast.java

```diff
@@ -0,0 +1,1413 @@
+        
+        mag01 = new int[2];
+        mag01[0] = 0x0;
+        mag01[1] = MATRIX_A;
+
+        mt[0]= (int)(seed & 0xffffffff);
+        for (mti=1; mti<N; mti++) 
+            {
+            mt[mti] = 
+                (1812433253 * (mt[mti-1] ^ (mt[mti-1] >>> 30)) + mti); 
+            /* See Knuth TAOCP Vol2. 3rd Ed. P.106 for multiplier. */
+            /* In the previous versions, MSBs of the seed affect   */
+            /* only MSBs of the array mt[].                        */
+            /* 2002/01/09 modified by Makoto Matsumoto             */
+            // mt[mti] &= 0xffffffff;
+            /* for >32 bit machines */
+            }
+        }
+
+
+    /**
+     * Sets the seed of the MersenneTwister using an array of integers.
+     * Your array must have a non-zero length.  Only the first 624 integers
+     * in the array are used; if the array is shorter than this then
+     * integers are repeatedly used in a wrap-around fashion.
+     */
+
+    synchronized public void setSeed(int[] array)
+        {
+        if (array.length == 0)
+            throw new IllegalArgumentException("Array length must be greater than zero");
+        int i, j, k;
+        setSeed(19650218);
+        i=1; j=0;
+        k = (N>array.length ? N : array.length);
+        for (; k!=0; k--) 
+            {
+            mt[i] = (mt[i] ^ ((mt[i-1] ^ (mt[i-1] >>> 30)) * 1664525)) + array[j] + j; /* non linear */
+            // mt[i] &= 0xffffffff; /* for WORDSIZE > 32 machines */
+            i++;
+            j++;
+            if (i>=N) { mt[0] = mt[N-1]; i=1; }
+            if (j>=array.length) j=0;
+            }
+        for (k=N-1; k!=0; k--) 
+            {
+            mt[i] = (mt[i] ^ ((mt[i-1] ^ (mt[i-1] >>> 30)) * 1566083941)) - i; /* non linear */
+            // mt[i] &= 0xffffffff; /* for WORDSIZE > 32 machines */
+            i++;
+            if (i>=N) 
+                {
+                mt[0] = mt[N-1]; i=1; 
+                }
+            }
+        mt[0] = 0x80000000; /* MSB is 1; assuring non-zero initial array */ 
+        }
+
+
+    public int nextInt()
+        {
+        int y;
+        
+        if (mti >= N)   // generate N words at one time
+            {
+            int kk;
+            final int[] mt = this.mt; // locals are slightly faster 
+            final int[] mag01 = this.mag01; // locals are slightly faster 
+            
+            for (kk = 0; kk < N - M; kk++)
+                {
+                y = (mt[kk] & UPPER_MASK) | (mt[kk+1] & LOWER_MASK);
+                mt[kk] = mt[kk+M] ^ (y >>> 1) ^ mag01[y & 0x1];
+                }
+            for (; kk < N-1; kk++)
+                {
+                y = (mt[kk] & UPPER_MASK) | (mt[kk+1] & LOWER_MASK);
+                mt[kk] = mt[kk+(M-N)] ^ (y >>> 1) ^ mag01[y & 0x1];
+                }
+            y = (mt[N-1] & UPPER_MASK) | (mt[0] & LOWER_MASK);
+            mt[N-1] = mt[M-1] ^ (y >>> 1) ^ mag01[y & 0x1];
+
+            mti = 0;
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 366:386, 403:423, 440:460, 476:496, 523:543, 568:588, 596:616, 633:653, 671:691, 709:729, 737:757, 783:803, 811:831, 852:872, 880:900, 958:978, 986:1006, 1014:1034, 1042:1062, 1095:1115, 1165:1185, 1201:1221.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.